packages/cursorless-vscode/src/ide/vscode/VSCodeScopeVisualizer/VscodeFancyRangeHighlighter/generateDecorationsForCharacterRange/handleMultipleLines.test.ts

```diff
@@ -0,0 +1,142 @@
+import assert = require("assert");
+import { BorderStyle } from "../decorationStyle.types";
+import { handleMultipleLines } from "./handleMultipleLines";
+import { Range } from "@cursorless/common";
+import { map } from "itertools";
+
+const solid = BorderStyle.solid;
+const porous = BorderStyle.porous;
+const none = BorderStyle.none;
+
+/** `[start, end]` */
+type CharacterOffsets = [number, number];
+
+/** `[[firstLineStart, firstLineEnd], ...restLineEnds]` */
+type Input = [CharacterOffsets, ...number[]];
+
+/** `[lineNumber, [start, end], [top, right, bottom, left]` */
+type LineDecorations = [
+  number,
+  CharacterOffsets,
+  [BorderStyle, BorderStyle, BorderStyle, BorderStyle],
+];
+
+interface TestCase {
+  /**
+   * The input to `handleMultipleLines`, in the format
+   *
+   * ```
+   * [[firstLineStart, firstLineEnd], ...restLineEnds]
+   * ```
+   *
+   * We use a single number for lines after the first because they always start
+   * at character 0.
+   *
+   * The first line will have line number 0, and the rest will count up from
+   * there.
+   */
+  input: Input;
+
+  /**
+   * Each entry in this array is a list of expected highlights for a single
+   * line, each in the format
+   *
+   * ```
+   * [lineNumber, [start, end], [top, right, bottom, left]
+   * ```
+   */
+  expected: LineDecorations[];
+}
+
+const testCases: TestCase[] = [
```

Comment:
This test representation is a bit terse, but the hover is really helpful for understanding the shape, and you get the hang of it pretty quickly.  Otherwise the tests become a bit annoying to read because every single attribute gets its own line so it's about 10x as many lines